### PR TITLE
[MXNET-1185] Support large array in several operators (part 1)

### PR DESCRIPTION
**Pull request by apeforest**
## Description ##
This PR fixed the large array issue (#13036, #13070) in the following operators:
ndarray.ones
ndarray.zeros
ndarray.sum
ndarray.slice
ndarray.random.uniform
ndarray.empty

This PR is only the first effort that addresses some limitations in some basic operators. More tests and fix are coming as we identify more operators. This is part of a large project tracked in JIRA by https://issues.apache.org/jira/browse/MXNET-1184
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Nightly tests are added for complicated/long-running ones
- [X] Code is well-documented: 
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [X] Support large array (> 3billion elements) in certain operators
- [X] Added nightly tests on arrays of 5 billion elements

## Comments ##
This PR only supports the total number of elements in the array to go beyond 2 billion. The max shape size in any of the dimension is still limited to 2^31 because this PR does not update the C API. Due to the large change required in all language bindings, we decided to break the original PR (https://github.com/apache/incubator-mxnet/pull/13191) into several smaller parts. And this PR is part 1. 

This PR only changes the indexing variable type from int32_t to int64_t, not the data type of each element. No performance impact observed. For future guidance of data types to choose, please refer to https://cwiki.apache.org/confluence/display/MXNET/Large+Tensor+Support

### CODE REVIEW

**Comment by zheng-da** [comment]
The PR looks good to me.

**Comment by zheng-da** [comment]
thanks for the great work. your code modifies a lot of code and the test doesn't cover all of them.
could you please add more unit tests to make sure the modified code is covered?
Here I tried a bunch of tests. Most of them work. Unfortunately, broadcast div fails. Could you please double check the problem in broadcast div?

```python
import mxnet as mx
import numpy as np
long_dim = 50000000
arr = mx.nd.ones(shape=(long_dim, 200))

deg = mx.nd.ones(shape=(long_dim,1)) * 2
# Test broadcast div
res = arr/deg
assert np.sum(arr[-1].asnumpy() == 0.5) == arr.shape[1]

# Test element-wise
arr2 = mx.nd.ones(shape=(long_dim, 200))
res = arr + arr2
assert np.sum(res[-1].asnumpy() == 2) == arr.shape[1]
res = arr + 1
assert np.sum(res[-1].asnumpy() == 2) == arr.shape[1]
res = mx.nd.sqrt(arr + 3)
assert np.sum(res[-1].asnumpy() == 2) == arr.shape[1]

# Test reduce
assert mx.nd.sum(arr).asnumpy() == arr.shape[0] * arr.shape[1]

# Test dot
weight = mx.nd.ones(shape=(200, 100))
res = mx.nd.dot(arr, weight)
assert np.sum(res[-1].asnumpy() == 200) == weight.shape[1]

# Test FullyConnected
res = mx.nd.FullyConnected(arr, weight, num_hidden=weight.shape[1], no_bias=True)
assert np.sum(res[-1].asnumpy() == 200) == weight.shape[1]

# Test broadcast
range = mx.nd.arange(0, long_dim).reshape(long_dim, 1)
res = mx.nd.broadcast_to(range, shape=(range.shape[0], 200))
assert np.sum(res[-1].asnumpy() == long_dim) == res.shape[1]
res = mx.nd.broadcast_like(range, arr)
assert np.sum(res[-1].asnumpy() == long_dim) == arr.shape[1]

# Test clip
data = res
res = mx.nd.clip(data, a_min=100, a_max=1000)
assert np.sum(res[-1].asnumpy() == 1000) == arr.shape[1]


# Test take
idx = mx.nd.arange(long_dim-1000, long_dim)
res = mx.nd.take(arr, idx)
assert np.sum(res[-1].asnumpy() == 1) == res.shape[1]

# Test slice
res = mx.nd.slice(arr, begin=(long_dim-1000, 1), end=(long_dim, 100))
assert np.sum(res[-1].asnumpy() == 1) == res.shape[1]

# Test slice assign
res = arr.copy()
res[long_dim-1:long_dim] = 1000
assert np.sum(res[-1].asnumpy() == 1000) == arr.shape[1]

# Test expand_dims
res = mx.nd.expand_dims(arr, axis=1)
assert res.shape == (arr.shape[0], 1, arr.shape[1])

# Test squeeze
data = res
res = mx.nd.squeeze(data)
assert sum(res.shape == arr.shape) == 2

```

**Comment by vandanavk** [comment]
@mxnet-label-bot add [pr-awaiting-review, NDArray]

**Comment by apeforest** [comment]
@zheng-da I have fixed the bug in broadcast_div and added the tests you provided. Thanks for your review and testing!

File: tests/nightly/test_large_array.py [hunk 1]
```diff
@@ -15,20 +15,118 @@
...
+    assert np.sum(res[-1].asnumpy() == 1) == res.shape[1]
+
+def test_slice():
+    a = nd.ones(shape=(LARGE_X, SMALL_Y))
+    res = nd.slice(a, begin=(LARGE_X-1000, 1), end=(LARGE_X, SMALL_Y))
+    assert np.sum(res[-1].asnumpy() == 1) == res.shape[1]
+
+def test_slice_assign():
+    a = nd.ones(shape=(LARGE_X, SMALL_Y))
+    res = a.copy()
```
[the remark this answers is not in the document]
Comment (apeforest) [reply tests/nightly/test_large_array.py]: done

File: tests/nightly/test_large_array.py [hunk 2]
```diff
@@ -15,20 +15,118 @@
...
+def test_broadcast():
+    a = nd.ones(shape=(LARGE_X, SMALL_Y))
+    b = nd.arange(0, LARGE_X).reshape(LARGE_X, 1)
+    res = nd.broadcast_to(b, shape=(b.shape[0], SMALL_Y))
+    assert np.sum(res[-1].asnumpy() == LARGE_X) == res.shape[1]
+    res = mx.nd.broadcast_like(b, a)
+    assert np.sum(res[-1].asnumpy() == LARGE_X) == a.shape[1]
+
+def test_clip():
+    a = nd.arange(0, LARGE_Y).reshape(LARGE_Y, 1)
```
[the remark this answers is not in the document]
Comment (apeforest) [reply tests/nightly/test_large_array.py]: done

**Comment by zheng-da** [comment]
Could you fix this bug as well? thanks.

```python
def run_linear():
    data = mx.nd.ones(shape=(50*1000*1000, 100))
    linear = gluon.nn.Dense(100)
    linear.initialize(ctx=mx.cpu(0))
    res = linear(data)
    res.wait_to_read()
```

**Comment by apeforest** [comment]
@zheng-da I don't see problem with this test. I have added it to the nightly test script. Please review.

**Comment by zheng-da** [comment]
@apeforest i can't run the test for the dense layer in my C5.18x.

**Comment by zheng-da** [comment]
do you build your code with mkldnn or without mkldnn?

**Comment by zheng-da** [comment]
i can verify that the test code for dense works when mxnet isn't built with mkldnn.

**Comment by anirudh2290** [comment]
@pengzhao-intel raised some good concerns on #13036 about performance. Also from @wkcn tests  I see div op performance affected between int32 and int64. Have we done any micro-benchmarks for the operators affected ?

File: tests/nightly/test_large_array.py [hunk 3]
```diff
@@ -15,20 +15,126 @@
...
-        ind = nd.zeros((700000, 128))
-        x = m(ind)
-        x.shape
-        test = x.asnumpy()
-        assert (x.shape == test.shape)
 
 if __name__ == '__main__':
-    unittest.main()
+    import nose
+    nose.runmodule()
```
Comment (lebeg) [inline tests/nightly/test_large_array.py]: Empty line at the end

Comment (apeforest) [reply tests/nightly/test_large_array.py]: added

**Comment by apeforest** [comment]
@anirudh2290 This only changes the index type from int32_t to int64_t, not the data type of each element. So the performance impact is different from @wkcn test.

I also did a test in broadcast_div operator in MXNet and found no obvious runtime change (int64_t version actually is slightly faster)

`import mxnet as mx`
`import numpy as np`
`from mxnet.test_utils import check_speed`

`a = mx.sym.ones(shape=(100000, 10000))`
`b = mx.sym.ones(shape=(100000, 1)) * 2`
`c = mx.sym.broadcast_div(a, b)`

`print(check_speed(c))`

Master branch: 15.357867193222045 seconds
PR branch: 12.930877685546875 seconds

**Comment by apeforest** [comment]
@anirudh2290 Given the overhead of repeating CI, can we get this merged and add unit test for scatter_nd in an second PR? This PR is only part-1 of a larger PR anyways. Thanks.

**Comment by pengzhao-intel** [comment]
> @pengzhao-intel raised some good concerns on #13036 about performance. Also from @wkcn tests I see div op performance affected between int32 and int64. Have we done any micro-benchmarks for the operators affected ?

@anirudh2290  Agree with you. 
I see the changes make the INT64 by default in several OPs. 
Is it a little aggresive? How about use template to switch the data type?

**Comment by zheng-da** [comment]
@pengzhao-intel I think the performance impact is minimal. Only the index type is changed. The operations on the index is basically add and comparison. All other computation is much more expensive. I double there is measurable performance difference.

**Comment by pengzhao-intel** [comment]
@zheng-da @apeforest  I am fine with the current PR since it's only the first part of the changes. 

I have re-visited the proposal [in here](https://cwiki.apache.org/confluence/display/MXNET/Large+Tensor+Support) and I suggest to add more details and data, such as how CUDA and MKL-DNN supports INT64 type and any potential impacts.

**Comment by apeforest** [comment]
Thank god the CI tests finally passed!
@anirudh2290 A JIRA epic has already been created to track this project: https://issues.apache.org/jira/browse/MXNET-1184. I will bring this proposal to dev list for discusssion. Since this change should update APIs in all language bindings, help is needed from experts in different language support.

**Pull request merged**